File: server/src/services/paymentService.js
import { getRazorpayInstance } from "../config/razorpay.js";
import { envConfig } from "../config/env.js";
import { paymentRepository } from "../repositories/paymentRepository.js";
import { CryptoUtils } from "../utils/cryptoUtils.js";
import { BadRequestError, NotFoundError, PaymentVerificationError } from "../utils/AppError.js";
import { logger } from "../utils/logger.js";

/**
 * Service Layer: Payment Business Logic Orchestrator.
 * Connects controller requests with external Razorpay SDK and database storage.
 */
export class PaymentService {
  /**
   * Returns public Razorpay Key ID
   */
  getPublicApiKey() {
    if (!envConfig.RAZORPAY_KEY_ID) {
      throw new BadRequestError("Razorpay Key ID is not configured on the server.");
    }
    return { key: envConfig.RAZORPAY_KEY_ID };
  }

  /**
   * Create Razorpay Order and persist initial transaction in DB
   * @param {number} amount - Amount in INR
   * @param {string} currency - Currency code (default INR)
   */
  async createCheckoutOrder(amount, currency = "INR") {
    if (!amount || isNaN(amount) || amount < 1) {
      throw new BadRequestError("Invalid checkout amount. Amount must be a valid number >= 1 INR.");
    }

    const razorpay = getRazorpayInstance();
    const receiptId = `receipt_${Date.now()}_${Math.floor(Math.random() * 1000)}`;

    const options = {
      amount: Math.round(amount * 100), // Razorpay accepts amount in sub-units (paise)
      currency,
      receipt: receiptId,
    };

    logger.info(`Initiating Razorpay order creation for amount: ${amount} ${currency}`);
    const order = await razorpay.orders.create(options);

    // Save initial transaction state in DB
    const paymentRecord = await paymentRepository.create({
      razorpayOrderId: order.id,
      amount,
      currency,
      receipt: receiptId,
      status: "created",
    });

    logger.info(`Checkout order created successfully. Order ID: ${order.id}`);
    return { order, paymentId: paymentRecord._id };
  }

  /**
   * Verify Razorpay Payment Signature and update transaction status
   */
  async verifyPayment(orderId, paymentId, signature) {
    if (!orderId || !paymentId || !signature) {
      throw new BadRequestError("Missing required payment verification credentials.");
    }

    const secret = envConfig.RAZORPAY_KEY_SECRET;
    const isValid = CryptoUtils.verifyPaymentSignature(orderId, paymentId, signature, secret);

    const paymentRecord = await paymentRepository.findByOrderId(orderId);

    if (!isValid) {
      logger.warn(`Payment signature verification failed for Order ID: ${orderId}`);
      if (paymentRecord) {
        await paymentRepository.updateStatus(orderId, {
          status: "failed",
          failureReason: "Invalid HMAC signature",
        });
      }
      throw new PaymentVerificationError("Invalid cryptographic payment signature.");
    }

    logger.info(`Payment signature verified successfully for Order ID: ${orderId}`);

    const updatedPayment = await paymentRepository.updateStatus(orderId, {
      razorpayPaymentId: paymentId,
      razorpaySignature: signature,
      status: "completed",
    });

    return {
      success: true,
      paymentId,
      orderId,
      status: "completed",
      updatedRecord: updatedPayment,
    };
  }

  /**
   * Retrieve payment transaction logs with pagination
   */
  async getPaymentHistory(page = 1, limit = 20) {
    return await paymentRepository.findAll(page, limit);
  }
}

export const paymentService = new PaymentService();

File: dotnet-server/Services/PaymentService.cs
using Microsoft.Extensions.Options;
using Razorpay.Api;
using PaymentGateway.Api.Config;
using PaymentGateway.Api.DTOs;
using PaymentGateway.Api.Models;
using PaymentGateway.Api.Repositories;
using PaymentGateway.Api.Utilities;

namespace PaymentGateway.Api.Services;

public interface IPaymentService
{
    string GetPublicKey();
    Task<object> CreateCheckoutOrderAsync(decimal amount, string currency);
    Task<object> VerifyPaymentAsync(VerifyPaymentRequestDto dto);
    Task<List<PaymentRecord>> GetPaymentHistoryAsync(int page, int limit);
}

public class PaymentService : IPaymentService
{
    private readonly RazorpaySettings _settings;
    private readonly IPaymentRepository _repository;
    private readonly ILogger<PaymentService> _logger;

    public PaymentService(
        IOptions<RazorpaySettings> settings,
        IPaymentRepository repository,
        ILogger<PaymentService> logger)
    {
        _settings = settings.Value;
        _repository = repository;
        _logger = logger;
    }

    public string GetPublicKey()
    {
        if (string.IsNullOrEmpty(_settings.KeyId))
        {
            throw new BadRequestException("Razorpay Key ID is not configured on server.");
        }
        return _settings.KeyId;
    }

    public async Task<object> CreateCheckoutOrderAsync(decimal amount, string currency)
    {
        if (amount < 1)
        {
            throw new BadRequestException("Amount must be at least 1 INR.");
        }

        var client = new RazorpayClient(_settings.KeyId, _settings.KeySecret);
        var receiptId = $"receipt_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_{Random.Shared.Next(100, 999)}";

        var input = new Dictionary<string, object>
        {
            { "amount", Convert.ToInt64(amount * 100) }, // Razorpay expects amount in paise
            { "currency", currency },
            { "receipt", receiptId }
        };

        _logger.LogInformation("Creating Razorpay Order for amount {Amount} {Currency}", (double)amount, currency);
        Order order = client.Order.Create(input);
        string orderId = Convert.ToString(order["id"]) ?? string.Empty;

        var record = new PaymentRecord
        {
            RazorpayOrderId = orderId,
            Amount = amount,
            Currency = currency,
            Receipt = receiptId,
            Status = "created",
            CreatedAt = DateTime.UtcNow
        };

        await _repository.CreateAsync(record);
        _logger.LogInformation("Successfully created Razorpay Order ID: {OrderId}", orderId);

        return new
        {
            order = new
            {
                id = orderId,
                amount = Convert.ToInt64(amount * 100),
                currency,
                receipt = receiptId
            }
        };
    }

    public async Task<object> VerifyPaymentAsync(VerifyPaymentRequestDto dto)
    {
        var isValid = CryptoUtils.VerifyPaymentSignature(
            dto.RazorpayOrderId,
            dto.RazorpayPaymentId,
            dto.RazorpaySignature,
            _settings.KeySecret
        );

        if (!isValid)
        {
            _logger.LogWarning("Payment signature verification failed for Order ID: {OrderId}", dto.RazorpayOrderId);
            await _repository.UpdateStatusAsync(dto.RazorpayOrderId, "failed", failureReason: "Invalid HMAC signature");
            throw new PaymentVerificationException("Invalid cryptographic payment signature.");
        }

        _logger.LogInformation("Payment signature verified successfully for Order ID: {OrderId}", dto.RazorpayOrderId);
        var updated = await _repository.UpdateStatusAsync(
            dto.RazorpayOrderId,
            "completed",
            dto.RazorpayPaymentId,
            dto.RazorpaySignature
        );

        return new
        {
            success = true,
            payment_id = dto.RazorpayPaymentId,
            order_id = dto.RazorpayOrderId
        };
    }

    public async Task<List<PaymentRecord>> GetPaymentHistoryAsync(int page, int limit)
    {
        return await _repository.GetAllAsync(page, limit);
    }
}
